rad/prelude/cmd-parsing.rad

```diff
@@ -27,32 +29,57 @@
 
-(def /cmd-1-opts
-  "Matches a command `cmd` with one argument and an optional
-  combination of options that are allowed by opts-map"
-  (fn [cmd opts opts-map]
+(def /cmd-opts
+  "Matches the command `cmd` and a combination of optional parameters, where
+  `opt-match` matches the options to the keys of `multi-ops`. `multi-ops` is a
+  dict of option-keys with multiple arguments. If no options are passed, the
+  `default` is returned."
+  (fn [cmd opts multi-opts opt-match default help]
+    (def-rec parse-cli
+      (fn [args default]
+        (def parse-arg
+          (fn [name args]
+            (match args
+              (/cons 'val 'rest)
+                  (if (elem? val (lookup name multi-opts))
+                    (modify-map name (fn [x] (cons (read (string-append ":" val)) x)) (parse-cli rest default))
+                    (parse-failure (string-append "Invalid argument " val " for option " (show name) ".") help))
+              []  (parse-failure (string-append "Missing argument for option " (show name) ".") help)
+              )))
+        (def parse-opt
+          (fn [opt rest]
+            (def option (opt-match opt))
+            (if (eq? (lookup-maybe option multi-opts) :nothing)
+              (insert option #t (parse-cli rest default))
+              (parse-arg option rest))))
+        (match args
+          /nil                default
+          (/cons 'opt 'rest)  (parse-opt opt rest))))
     (fn [v]
       (match v
-        [cmd]           [:just {opts []}]
-        (/cons cmd 'os) [:just {opts (opts-map os)}]
+        [cmd]           [:just {opts default}]
+        (/cons cmd 'os) [:just {opts (parse-cli os default)}]
         _               :nothing))))
 
-(:test "/cmd-1-opts"
+(:test "/cmd-opts"
   [:setup
     (do
-      (def opts-map
-        (fn [opts]
-          (map
-            (fn [opt]
-              (if (elem? opt ["--new" "--old"])
-                (drop 2 opt)
-                "unknown filter")) opts))))]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list"]) ==> [:just {opts []}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["no-list"]) ==> :nothing]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--old"]) ==> [:just {opts ["old"]}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--old" "--new"]) ==> [:just {opts ["old" "new"]}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--filter"]) ==> [:just {opts ["unknown filter"]}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--old" "--current"]) ==> [:just {opts ["old" "unknown filter"]}]])
+      (def help "help")
+      (def multi-opts {:state ["acc" "prop"]})
+      (def default {:state [] :fancy #f})
+      (def opt-match
+        (fn [opt]
+          (match opt
+            "--filter-by-state" :state
+            "--state"           :state
+            "-s"                :state
+            "--fancy"           :fancy
+            'invalid            (parse-failure (string-append "Invalid option " invalid " for command list.") help)))))]
```

Comment:
Instead of `opt-match` being a function it could be a dictionary. The user of `/cmd-opts` would not need to provide the `'invalid` case every time.